fili-core/src/main/java/com/yahoo/bard/webservice/data/config/table/PhysicalTableDefinition.java

```diff
@@ -1,93 +1,144 @@
+     * Get the set of physical tables required to build the current physical table.
+     *
+     * @return a set of physical table names
+     */
+    public abstract Set<TableName> getDependentTableNames();
+
+    /**
+     * Given a set of metric names and a data source metadata service, build the corresponding physical table.
      *
-     * @param name  The table name
-     * @param grain  The zoneless time grain
-     * @param dimensions  The dimension configurations
+     * @param dictionaries  Dictionary containing dimension dictionary and physical table dictionary
+     * @param metadataService  Service containing column available interval information
      *
-     * @deprecated The time zone of a physical table should be set explicitly rather than rely on defaulting to UTC
+     * @return optional of the corresponding type of physical table, empty optional if could not be build.
      */
-    @Deprecated
-    public PhysicalTableDefinition(
-            TableName name,
-            ZonelessTimeGrain grain,
-            Iterable<? extends DimensionConfig> dimensions
-    ) {
-        this(name, grain.buildZonedTimeGrain(DateTimeZone.UTC), dimensions);
-    }
+    public abstract Optional<PhysicalTable> build(
+            ResourceDictionaries dictionaries,
+            DataSourceMetadataService metadataService
+    );
 
     public TableName getName() {
         return name;
     }
 
-    public ZonedTimeGrain getGrain() {
-        return grain;
+    public ZonedTimeGrain getTimeGrain() {
+        return timeGrain;
     }
 
-    public ImmutableSet<? extends DimensionConfig> getDimensions() {
-        return dimensions;
+    public Set<FieldName> getMetricNames() {
+        return metricNames;
+    }
+
+    public Set<? extends DimensionConfig> getDimensionConfigs() {
+        return dimensionConfigs;
     }
 
     public Map<String, String> getLogicalToPhysicalNames() {
         return logicalToPhysicalNames;
     }
+
+    /**
+     * Builds the dimension logical name to physical name mapping from dimension configs.
+     *
+     * @param dimensionConfigs  Dimension config containing both logical and physical names
+     *
+     * @return the dimension logical name to physical name mapping
+     */
+    protected Map<String, String> buildLogicalToPhysicalNames(Set<? extends DimensionConfig> dimensionConfigs) {
+        return dimensionConfigs.stream()
+                .collect(
+                        Collectors.toMap(
+                                DimensionConfig::getApiName,
+                                config -> {
+                                    String physicalName = config.getPhysicalName();
+                                    if (physicalName == null) {
+                                        throw new RuntimeException("No physical name found for dimension: "
+                                                + config.getApiName());
+                                    }
```

Comment:
I'm not sure this behavior is correct, but I'm also not sure what the correct behavior _is_. 

Per the docs on `DimensionConfig::getPhysicalName`, this value: `This field (if set) is used as the only physical name`, which means that it is OK not to have a value for this field. It also means that if it's _not_ set, the expectation is that there is more than 1 physical name for this dimension, and that information is being set on the PhysicalTables through the mapping directly.

That said, if this method is being made available as a default / helper (ie. if no mapping was provided), which is the expected use of this I think, then there _should_ be a single physical name present for each dimension.

The other concern I have is that by throwing an exception, we're forcing users to specify the physical name of the dimension. While I think it's highly likely that most Fili instances will have different API names and physical names for dimensions, would it ease onboarding (ie. reduce required configuration) if we instead had this method log a warning, and then use the API name as the default physical name?